### PR TITLE
adapter: use STORAGE ReadHold instead of custom hold tracking in adapter

### PR DESCRIPTION
**Pull request by aljoscha**
Preparatory work for https://github.com/MaterializeInc/materialize/issues/24845, where we want to introduce more concurrency
to the Coordinator and Controllers.

Before, adapter would track outstanding `ReadHolds` in its own state and
continually update the `ReadPolicy` of storage collections based on a base
policy and outstanding read holds.

Now, we go directly to STORAGE for acquiring ephemeral `ReadHolds`. Note
that the adapter still keeps `TimelineReadHolds` for collections, which
control the `ReadPolicy` that we continually update. This is so that
timeline-dependent collections remain readable at the oracle read
timestamp even when there are no outstanding other (ephemeral) read
holds.

This leaves the COMPUTE part of `ReadHolds` untouched.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by aljoscha** [comment]
Things are starting to become more interesting now! @teskje you'll notice that the new `ReadHold` is not depending on any storage code, so I could move that to a common location if/when compute wants to use the same type.

File: src/storage-types/src/read_holds.rs
Comment (rjobanp) [inline src/storage-types/src/read_holds.rs:102]: shouldn't this say `current since is beyond new frontier`?

Comment (aljoscha) [reply src/storage-types/src/read_holds.rs:102]: absolutely right, that this is wrong! I think I was going for `new frontier ... is **NOT** beyond current since ...`

**Review by maddyblue — COMMENTED** [review]
Adapter parts lgtm.

File: src/storage-client/src/controller.rs
Comment (teskje) [inline src/storage-client/src/controller.rs:570]: I read this comment to mean that if I request a read hold before the `since`, I will instead receive one at the `since`? I think it would be more robust to instead return an error. This avoids bugs when the caller simply assumes that it got the requested read hold without checking.

Taking a step back, there are two scenarios I can think of when this API is called:
1. The caller wants a read hold at the earliest possible time, or just any read hold.
2. The caller wants a read hold at one specific time.

With this API, a caller can achieve (1) by `acquire_read_hold(id, Timestamp::MIN);` and (2) by `let hold = acquire_read_hold(id, requested_time); hold.try_downgrade(requested_time)?;`. But forgetting the check is easy.

How about we instead make the desired parameter an `Option` that when set to `Some` requests a precise read hold? Then the above usages would look like (1) `acquire_read_hold(id, None)` and (2) `acquire_read_hold(id, Some(requested_time))?`. Alternatively, two different methods, one that takes a desired frontier and one that doesn't would work as well. The latter could even be a wrapper around the former that just calls `try_downgrade`.

File: src/storage-types/src/read_holds.rs
Comment (teskje) [inline src/storage-types/src/read_holds.rs:31]: I think these fields shouldn't be `pub`?

File: src/storage-types/src/read_holds.rs
Comment (teskje) [inline src/storage-types/src/read_holds.rs:70]: The semantics of this method are a bit unclear. What happens if I give it two read holds for different `id`s? Looking at the implementation it looks like we will treat `other` like it is a read hold for the same collection as `self` and then we might end up with (a) a frontier regression for `self.id` and (b) a leaked read hold for `other.id`? Also weird things will happen when you get two read holds with different `holds_tx`s.

I think you'd at least want to assert that the two IDs and channels are equal. And then, isn't the implementation equivalent to just dropping the `ReadHold` with the greater `since`?

File: src/storage-controller/src/lib.rs
Comment (teskje) [inline src/storage-controller/src/lib.rs:1851]: I was confused by this variable name because I thought the channel would only contain `ReadHold::drop` events. But it also contains downgrades and clone events, so how about calling it `changes` instead?

File: src/storage-types/src/read_holds.rs
Comment (teskje) [inline src/storage-types/src/read_holds.rs:126]: Do you want to keep this log in? Might want to consider using `trace!` instead, depending on how often this will fire in an idle environment that receives various metrics queries periodically.

File: src/storage-types/src/read_holds.rs
Comment (teskje) [inline src/storage-types/src/read_holds.rs:165]: Nit: If you encapsulate the send in a method you don't have to always repeat that hanging up is ok :)

File: src/compute-client/src/controller/instance.rs
Comment (teskje) [inline src/compute-client/src/controller/instance.rs:1002]: Can you add a TODO here to store the read holds instead of installing read capabilities with the storage controller manually?

File: src/storage-client/src/controller.rs
Comment (aljoscha) [reply src/storage-client/src/controller.rs:570]: Yeah, I've thought about this in the context of the adapter version of `acquire_read_holds` before. After a couple of recent refactorings, there's actually zero places in the code where we call that with a timestamp other than `MIN` _or_ with `precise = true`. So we can clean at least that up, I believe. cc @ggevay who initially added the `precise` parameter.

I'll push a commit that makes this here API more obvious.

File: src/storage-types/src/read_holds.rs
Comment (aljoscha) [reply src/storage-types/src/read_holds.rs:31]: I had them pub because I was constructing these in tests. Later I added `new()` so can now remove them. 👌

File: src/storage-types/src/read_holds.rs
Comment (aljoscha) [reply src/storage-types/src/read_holds.rs:70]: Oh boy, good catch! This code is quite old by now, and has remnants from a time when I had a `ReadHold` that contained multiple IDs. Will fix!

> I think you'd at least want to assert that the two IDs and channels are equal. And then, isn't the implementation equivalent to just dropping the ReadHold with the greater since?

With our timestamp yes, but because the timestamp type is generic here we could have more complicated `sinces` with partial orderings.

File: src/storage-types/src/read_holds.rs
Comment (aljoscha) [reply src/storage-types/src/read_holds.rs:165]: ahh, this made me realize that the other comment in `clone()` is wrong. We _do_ care about being able to send. The code was correct, it's just that the comment was outdated.

File: src/storage-client/src/controller.rs
Comment (ggevay) [reply src/storage-client/src/controller.rs:570]: > there's actually zero places in the code where we call that with a timestamp other than `MIN` or with `precise = true`

I have to say I'm happy about this! Feel free to remove the `precise` stuff then!

**Comment by aljoscha** [comment]
@teskje I think I addressed all your comments, PTAL. I also rebased and made sure things are nice and tidy again.

File: src/storage-client/src/controller.rs
Comment (teskje) [inline src/storage-client/src/controller.rs:589]:
```suggestion
    /// [ReadHoldError::SinceViolation] when that is not possible.
```

File: src/storage-client/src/controller.rs
Comment (teskje) [reply src/storage-client/src/controller.rs:570]: I agree, this is nicer! Still I'd prefer if we could remove the pitfall of `StorageController::acquire_read_holds` silently returning a hold that is later than requested. Also because now the `acquire_read_holds` and `acquire_read_hold_at_time` behave inconsistently (only one of them erroring on a too-early desired time).

AFAICT the only place this method is called externally is in adapter's `acquire_read_hold`, where the requested read hold is at `Timestamp::MIN`. So how about we change this method to one that always returns the `since` read holds?

```rust
    fn acquire_read_holds(
        &mut self,
        desired_holds: Vec<GlobalId>,
    ) -> Result<Vec<ReadHold<Self::Timestamp>>, ReadHoldError>;
```

Or, alternatively, if you think we might need a batched variant of `acquire_read_hold_at_time` in the future, this method could return a `ReadHoldError::SinceViolation` like `acquire_read_hold_at_time` does.

File: src/storage-controller/src/lib.rs
Comment (teskje) [inline src/storage-controller/src/lib.rs:1759]: Is it actually safe to re-enqueue this way? What if you have two updates for the same read hold? E.g.:

* `(u1, [(100, -1), (200, +1)])`, `(u1, [(200, -1), (300, +1)])`

After this little re-enqueuing stunt, the channel has the following contents:

* `(u1, [(200, -1), (300, +1)])`, `(u1, [(100, -1), (200, +1)])`

I _think_ that's fine based on how the consumer of `read_holds_rx` is implemented, it collects all queued changes first, and only then applies them. But it would be good to have a comment somewhere that this consumer behavior is important for correctness and not just a performance optimization.

File: src/storage-types/src/read_holds.rs
Comment (teskje) [reply src/storage-types/src/read_holds.rs:71]: Ah, good point!

**Review by teskje — APPROVED** [review]
LGTM. I think the `acquire_read_holds` API can be more foolproof, but that doesn't need to block merging.

File: src/storage-controller/src/lib.rs
Comment (aljoscha) [reply src/storage-controller/src/lib.rs:1759]: Yeah, I don't like this stunt, and newer versions of tokio actually have a `len()` and `is_empty()` on channels. I'll add comments that explain what's going on (which you already correctly intuited).

File: src/storage-types/src/read_holds.rs
Comment (aljoscha) [reply src/storage-types/src/read_holds.rs:70]: 👌

File: src/compute-client/src/controller/instance.rs
Comment (aljoscha) [reply src/compute-client/src/controller/instance.rs:1019]: man, I had thought about it but then didn't want to futz with your error definitions too much... will do! 😅

**Comment by aljoscha** [comment]
@teskje thanks for the thorough reviews! I think I managed to address all your comments, even if they were just nits. 😅

**Review by jkosh44 — APPROVED** [review]

File: src/storage-client/src/controller.rs
Comment (sploiselle) [reply src/storage-client/src/controller.rs:560]: I was a little confused as to the purpose of these read holds' ephemeral nature; took me a bit to understand that this is essentially trying to create a cross-sub-system txn where we:
1. Install an ephemeral read hold in the storage controller.
2. Allow the compute controller to instead the read hold at the ephemeral read hold's since
3. Release the ephemeral read hold, knowing that the compute controller has gotten what it needs.

Is that right? If so, I'd love to see this motivating use more explicitly suggested in either the comments, module docs, etc. I see that it's mentioned when you `drop` these read holds, but it was unclear to me that I would find the answer  to "why do we need these to be ephemeral?" exactly there and nowhere else. It seems like a fundamental enough tenet of their design that it could be brought up somewhere.

**Review by sploiselle — APPROVED** [review]
I have some questions/suggestions about documention/naming that I don't think are high nail. Would greatly appreciate being addressed before merging, but there's nothing technical wrong afaict so doesn't need to be blocking.

File: src/storage-types/src/read_holds.rs
Comment (teskje) [reply src/storage-types/src/read_holds.rs:24]: We plan to use this type throughout the controllers and their clients, so having a more handy name would be nice. I like `ReadHold`, but if there are concerns about ambiguity, how about `ReadCapability`? That'd match the name that Timely uses for the `Capability` type, which works similarly.

**Comment by aljoscha** [comment]
@teskje & @sploiselle Thanks for the reviews so far! I'll try and address here because related comments are now spread across multiple GH comments:

About naming and purpose: sorry if that was confusing, but the new `ReadHolds` are not meant to be ephemeral/transient. Eventually, the compute controller should only use this API instead of the lower-level `update_read_capabilities`. The neat thing is that the adapter can acquire the read holds and hand them off to compute, at which point compute doesn't even need to acquire it's own holds anymore. Plus, in my next commit/PR (see in-progress https://github.com/MaterializeInc/materialize/pull/24816), where I factor a `StorageCollections` out of the `StorageController`. The now-slimmer `StorageController` has `ReadHold`s from `StorageCollections`, to keep sinces where they need to be for ingestions.

It's probably easier to talk about this in person! 😅

**Comment by sploiselle** [comment]
@aljoscha I was unaware of the longer-term plans here, which I think make my comments moot. I'd like to see a mention of these plans in the code, lest something get derailed and the next generation of monkeys never learn what obelisk we were cruising toward. So, this doubly LGTM.

File: src/storage-client/src/controller.rs
Comment (sploiselle) [reply src/storage-client/src/controller.rs:560]: clarified here: https://github.com/MaterializeInc/materialize/pull/26700#issuecomment-2089204947

File: src/storage-types/src/read_holds.rs
Comment (sploiselle) [reply src/storage-types/src/read_holds.rs:24]: clarified here: https://github.com/MaterializeInc/materialize/pull/26700#issuecomment-2089204947

File: src/storage-controller/src/lib.rs
Comment (aljoscha) [reply src/storage-controller/src/lib.rs:1751]: will add this!

**Comment by aljoscha** [comment]
I merged. Thanks for all the reviews! ☺️

**Pull request merged**
